File: spark-events-ui/src/app/model/TaskMetrics.ts

/**
 * 
 */
export class InputMetrics {
  // @JsonProperty("Bytes Read")
  bytesRead: number;

  // @JsonProperty("Records Read")
  recordsRead: number;

  constructor( bytesRead: number, recordsRead: number ) {
    this.bytesRead = bytesRead;
    this.recordsRead = recordsRead;
  }

  static fromJson( src: any ): InputMetrics {
    let bytesRead = <number>src['Bytes read'];
    let recordsRead = <number>src['Records Read'];
    return new InputMetrics( bytesRead, recordsRead );
  }

  static createDefault(): InputMetrics {
    return new InputMetrics(0, 0);
  }

}

/**
 * 
 */
export class OutputMetrics {
  // @JsonProperty("Bytes Written")
  bytesWritten: number;

  // @JsonProperty("Records Written")
  recordsWritten: number;

  constructor( bytesWritten: number, recordsWritten: number ) {
    this.bytesWritten = bytesWritten;
    this.recordsWritten = recordsWritten;
  }

  static fromJson( src: any ): OutputMetrics {
    let bytesWritten = <number>src['Bytes Written'];
    let recordsWritten = <number>src['Records Written'];
    return new OutputMetrics( bytesWritten, recordsWritten );
  }

  static createDefault(): OutputMetrics {
    return new OutputMetrics(0, 0);
  }

}



/* 
 *
 */
export class ShuffleReadMetrics {
  // "Remote Blocks Fetched"
  remoteBlocksFetched: number;
  // "Local Blocks Fetched"
  localBlocksFetched: number;
  // "Fetch Wait Time"
  fetchWaitTime: number;
  // "Remote Bytes Read"
  remoteBytesRead: number;
  // "Remote Bytes Read To Disk"
  remoteBytesReadToDisk: number;
  // "Local Bytes Read"
  localBytesRead: number;
  // "Total Records Read"
  totalRecordsRead: number;

  constructor( remoteBlocksFetched: number, localBlocksFetched: number, fetchWaitTime: number,
    remoteBytesRead: number, remoteBytesReadToDisk: number, localBytesRead: number, totalRecordsRead: number ) {
    this.remoteBlocksFetched = remoteBlocksFetched;
    this.localBlocksFetched = localBlocksFetched;
    this.fetchWaitTime = fetchWaitTime;
    this.remoteBytesRead = remoteBytesRead;
    this.remoteBytesReadToDisk = remoteBytesReadToDisk;
    this.localBytesRead = localBytesRead;
    this.totalRecordsRead = totalRecordsRead;
  }

  static fromJson( src: any ): ShuffleReadMetrics {
    let remoteBlocksFetched = <number>src['Remote Blocks Fetched'];
    let localBlocksFetched = <number>src['Local Blocks Fetched'];
    let fetchWaitTime = <number>src['Fetch Wait Time'];
    let remoteBytesRead = <number>src['Remote Bytes Read'];
    let remoteBytesReadToDisk = <number>src['Remote Bytes Read To Disk'];
    let localBytesRead = <number>src['Local Bytes Read'];
    let totalRecordsRead = <number>src['Total Records Read'];
    return new ShuffleReadMetrics( remoteBlocksFetched, localBlocksFetched, fetchWaitTime, remoteBytesRead,
      remoteBytesReadToDisk, localBytesRead, totalRecordsRead );
  }

  static createDefault(): ShuffleReadMetrics {
    return new ShuffleReadMetrics(0, 0, 0, 0, 0, 0, 0);
  }
}


/**
 * 
 */
export class ShuffleWriteMetrics {

  //@JsonProperty("Shuffle Bytes Written")
  shuffleBytesWritten: number;

  //@JsonProperty("Shuffle Write Time")
  shuffleWriteTime: number;

  // @JsonProperty("Shuffle Records Written")
  shuffleRecordsWritten: number;

  constructor( shuffleBytesWritten: number, shuffleWriteTime: number, shuffleRecordsWritten: number ) {
    this.shuffleBytesWritten = shuffleBytesWritten;
    this.shuffleWriteTime = shuffleWriteTime;
    this.shuffleRecordsWritten = shuffleRecordsWritten;
  }

  static fromJson( src: any ): ShuffleWriteMetrics {
    let shuffleBytesWritten = <number>src['Shuffle Bytes Written'];
    let shuffleWriteTime = <number>src['Shuffle Write Time'];
    let shuffleRecordsWritten = <number>src['Shuffle Records Written'];
    return new ShuffleWriteMetrics( shuffleBytesWritten, shuffleWriteTime, shuffleRecordsWritten );
  }

  static createDefault(): ShuffleWriteMetrics {
    return new ShuffleWriteMetrics(0, 0, 0);
  }

}

/**
 * 
 */
export class UpdatedBlock extends Map<String, String> {
  // TODO
  static fromJson( src: any ): UpdatedBlock {
    return new UpdatedBlock( Object.entries( src ) );
  }

  static fromJsonArray( src: any[] ): UpdatedBlock[] {
    return src.map( x => UpdatedBlock.fromJson( x ) );
  }

  static createDefault(): UpdatedBlock {
    return new UpdatedBlock();
  }

}


/**
 * 
 */
export class TaskMetrics {

  // @JsonProperty("Executor Deserialize Time")
  executorDeseralizeTime: number;

  // @JsonProperty("Executor Deserialize CPU Time")
  executorDeseralizeCpuTime: number;

  // @JsonProperty("Executor Run Time")
  executorRunTime: number;

  // @JsonProperty("Executor CPU Time")
  executorCpuTime: number;

  // @JsonProperty("Peak Execution Memory")
  peakExecutionMemory: number;

  // @JsonProperty("Result Size")
  resultSize: number;

  // @JsonProperty("JVM GC Time")
  jvmGCTime: number;

  // @JsonProperty("Result Serialization Time")
  resultSerializationTime: number;

  // @JsonProperty("Memory Bytes Spilled")
  memoryBytesSpilled: number;

  // @JsonProperty("Disk Bytes Spilled")
  diskBytesSpilled: number;

  // @JsonProperty("Shuffle Read Metrics")
  shuffleReadMetrics: ShuffleReadMetrics;

  // @JsonProperty("Shuffle Write Metrics")
  shuffleWriteMetrics: ShuffleWriteMetrics;

  // @JsonProperty("Input Metrics")
  inputMetrics: InputMetrics;

  // @JsonProperty("Output Metrics")
  outputMetrics: OutputMetrics;

  // @JsonProperty("Updated Blocks")
  updatedBlocks: UpdatedBlock[];

  constructor(
    executorDeseralizeTime: number,
    executorDeseralizeCpuTime: number,
    executorRunTime: number,
    executorCpuTime: number,
    peakExecutionMemory: number,
    resultSize: number,
    jvmGCTime: number,
    resultSerializationTime: number,
    memoryBytesSpilled: number,
    diskBytesSpilled: number,
    shuffleReadMetrics: ShuffleReadMetrics,
    shuffleWriteMetrics: ShuffleWriteMetrics,
    inputMetrics: InputMetrics,
    outputMetrics: OutputMetrics,
    updatedBlocks: UpdatedBlock[]
  ) {
    this.executorDeseralizeTime = executorDeseralizeTime;
    this.executorDeseralizeCpuTime = executorDeseralizeCpuTime;
    this.executorRunTime = executorRunTime;
    this.executorCpuTime = executorCpuTime;
    this.peakExecutionMemory = peakExecutionMemory;
    this.resultSize = resultSize;
    this.jvmGCTime = jvmGCTime;
    this.resultSerializationTime = resultSerializationTime;
    this.memoryBytesSpilled = memoryBytesSpilled;
    this.diskBytesSpilled = diskBytesSpilled;
    this.shuffleReadMetrics = shuffleReadMetrics;
    this.shuffleWriteMetrics = shuffleWriteMetrics;
    this.inputMetrics = inputMetrics;
    this.outputMetrics = outputMetrics;
    this.updatedBlocks = updatedBlocks;
  }

  static fromJson( src: any ): TaskMetrics {
    let executorDeseralizeTime = <number>src['Executor Deserialize Time'];
    let executorDeseralizeCpuTime = <number>src['Executor Deserialize CPU Time'];
    let executorRunTime = <number>src['Executor Run Time'];
    let executorCpuTime = <number>src['Executor CPU Time'];
    let peakExecutionMemory = <number>src['Peak Execution Memory'];
    let resultSize = <number>src['Result Size'];
    let jvmGCTime = <number>src['JVM GC Time'];
    let resultSerializationTime = <number>src['Result Serialization Time'];
    let memoryBytesSpilled = <number>src['Memory Bytes Spilled'];
    let diskBytesSpilled = <number>src['Disk Bytes Spilled'];
    let shuffleReadMetricsObj = src['Shuffle Read Metrics'];
    let shuffleReadMetrics = ( shuffleReadMetricsObj ) ? ShuffleReadMetrics.fromJson( shuffleReadMetricsObj ) : ShuffleReadMetrics.createDefault();
    let shuffleWriteMetricsObj = src['Shuffle Write Metrics'];
    let shuffleWriteMetrics = ( shuffleWriteMetricsObj ) ? ShuffleWriteMetrics.fromJson( shuffleWriteMetricsObj ) : ShuffleWriteMetrics.createDefault();
    let inputMetrics = InputMetrics.fromJson( src['Input Metrics'] );
    let outputMetrics = OutputMetrics.fromJson( src['Output Metrics'] );
    let updatedBlocks = UpdatedBlock.fromJsonArray( <any[]>src['Updated Blocks'] );

    return new TaskMetrics(executorDeseralizeTime,
      executorDeseralizeCpuTime,
      executorRunTime,
      executorCpuTime,
      peakExecutionMemory,
      resultSize,
      jvmGCTime,
      resultSerializationTime,
      memoryBytesSpilled,
      diskBytesSpilled,
      shuffleReadMetrics!!,
      shuffleWriteMetrics!!,
      inputMetrics,
      outputMetrics,
      updatedBlocks);
  }

  static createDefault(): TaskMetrics {
    let executorDeseralizeTime = 0;
    let executorDeseralizeCpuTime = 0;
    let executorRunTime = 0;
    let executorCpuTime = 0;
    let peakExecutionMemory = 0;
    let resultSize = 0;
    let jvmGCTime = 0;
    let resultSerializationTime = 0;
    let memoryBytesSpilled = 0;
    let diskBytesSpilled = 0;
    let shuffleReadMetrics = ShuffleReadMetrics.createDefault();
    let shuffleWriteMetrics = ShuffleWriteMetrics.createDefault();
    let inputMetrics = InputMetrics.createDefault();
    let outputMetrics = OutputMetrics.createDefault();
    let updatedBlocks: UpdatedBlock[] = [];
    return new TaskMetrics(executorDeseralizeTime,
        executorDeseralizeCpuTime,
        executorRunTime,
        executorCpuTime,
        peakExecutionMemory,
        resultSize,
        jvmGCTime,
        resultSerializationTime,
        memoryBytesSpilled,
        diskBytesSpilled,
        shuffleReadMetrics,
        shuffleWriteMetrics,
        inputMetrics,
        outputMetrics,
        updatedBlocks);
  }
}

File: spark-event-gateway-srv/src/main/java/fr/an/spark/gateway/eventlog/model/TaskMetrics.java
package fr.an.spark.gateway.eventlog.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.val;

import java.util.Collection;
import java.util.List;

import static fr.an.spark.gateway.eventlog.model.InternalAccumulator.UPDATED_BLOCK_STATUSES;

@NoArgsConstructor @AllArgsConstructor
public class TaskMetrics {

    @JsonProperty("Executor Deserialize Time")
    public long executorDeserializeTime;
    
    @JsonProperty("Executor Deserialize CPU Time")
    public long executorDeserializeCpuTime;

    @JsonProperty("Executor Run Time")
    public long executorRunTime;
    
    @JsonProperty("Executor CPU Time")
    public long executorCpuTime;
                            
    @JsonProperty("Result Size")
    public long resultSize;
                           
    @JsonProperty("JVM GC Time")
    public long jvmGcTime;
                          
    @JsonProperty("Result Serialization Time")
    public long resultSerializationTime;

    @JsonProperty("Memory Bytes Spilled")
    public long memoryBytesSpilled;
                  
    @JsonProperty("Disk Bytes Spilled")
    public long diskBytesSpilled;

    @JsonProperty("Peak Execution Memory")
    public long peakExecutionMemory;

    @JsonProperty("Input Metrics")
    public InputMetrics inputMetrics = new InputMetrics();

    @JsonProperty("Output Metrics")
    public OutputMetrics outputMetrics = new OutputMetrics();

    @JsonProperty("Shuffle Read Metrics")
    public ShuffleReadMetrics shuffleReadMetrics = new ShuffleReadMetrics();

    @JsonProperty("Shuffle Write Metrics")
    public ShuffleWriteMetrics shuffleWriteMetrics = new ShuffleWriteMetrics();

    @JsonProperty("Updated Blocks")
    public List<BlockIdStatus> updatedBlockStatuses;

    // -----------------------------------------------------------------------------------------------------------------

    public void setUpdatedBlockStatuses(List<BlockIdStatus> value) {
        this.updatedBlockStatuses = value;
    }

    public void incrMetrics(TaskMetrics other) {
        this.executorDeserializeTime += other.executorDeserializeTime;
        this.executorDeserializeCpuTime += other.executorDeserializeCpuTime;
        this.executorRunTime += other.executorRunTime;
        this.executorCpuTime += other.executorCpuTime;
        this.resultSize += other.resultSize;
        this.jvmGcTime += other.jvmGcTime;
        this.resultSerializationTime += other.resultSerializationTime;
        this.memoryBytesSpilled += other.memoryBytesSpilled;
        this.diskBytesSpilled += other.diskBytesSpilled;
        this.peakExecutionMemory = Math.max(this.peakExecutionMemory, other.peakExecutionMemory);
        if (other.inputMetrics != null) {
            if (this.inputMetrics == null) {
                this.inputMetrics = new InputMetrics();
            }
            inputMetrics.incrMetrics(other.inputMetrics);
        }
        if (other.outputMetrics != null) {
            if (this.outputMetrics == null) {
                this.outputMetrics = new OutputMetrics();
            }
            outputMetrics.incrMetrics(other.outputMetrics);
        }
        if (other.shuffleReadMetrics != null) {
            if (this.shuffleReadMetrics == null) {
                this.shuffleReadMetrics = new ShuffleReadMetrics();
            }
            shuffleReadMetrics.incrMetrics(other.shuffleReadMetrics);
        }
        if (other.shuffleWriteMetrics != null) {
            if (this.shuffleWriteMetrics == null) {
                this.shuffleWriteMetrics = new ShuffleWriteMetrics();
            }
            shuffleWriteMetrics.incrMetrics(other.shuffleWriteMetrics);
        }
        // ?? updatedBlockStatuses;
    }

    // -----------------------------------------------------------------------------------------------------------------

    @NoArgsConstructor @AllArgsConstructor
    public static class InputMetrics {
        @JsonProperty("Bytes Read")
        public long bytesRead;

        @JsonProperty("Records Read")
        public long recordsRead;

        public void incrMetrics(InputMetrics other) {
            this.bytesRead += other.bytesRead;
            this.recordsRead += other.recordsRead;
        }
    }

    @NoArgsConstructor @AllArgsConstructor
    public static class OutputMetrics {
        @JsonProperty("Bytes Written")
        public long bytesWritten;

        @JsonProperty("Records Written")
        public long recordsWritten;

        public void incrMetrics(OutputMetrics other) {
            this.bytesWritten += other.bytesWritten;
            this.recordsWritten += other.recordsWritten;
        }
    }



    @NoArgsConstructor @AllArgsConstructor
    public static class ShuffleReadMetrics {
	    @JsonProperty("Remote Blocks Fetched")
	    public long remoteBlocksFetched;
	    
	    @JsonProperty("Local Blocks Fetched")
	    public long localBlocksFetched;
	    
	    @JsonProperty("Fetch Wait Time")
	    public long fetchWaitTime;
	    
	    @JsonProperty("Remote Bytes Read")
	    public long remoteBytesRead;
	    
	    @JsonProperty("Remote Bytes Read To Disk")
	    public long remoteBytesReadToDisk;
	    
	    @JsonProperty("Local Bytes Read")
	    public long localBytesRead;
	    
	    @JsonProperty("Total Records Read")
	    public long recordsRead;

        @JsonProperty("Remote Requests Duration")
        public long remoteReqsDuration;

        @JsonProperty("Push Based Shuffle")
        public ShufflePushReadMetrics shufflePushReadMetrics;

        public void incrMetrics(ShuffleReadMetrics other) {
            this.remoteBlocksFetched += other.remoteBlocksFetched;
            this.localBlocksFetched += other.localBlocksFetched;
            this.fetchWaitTime += other.fetchWaitTime;
            this.remoteBytesRead += other.remoteBytesRead;
            this.remoteBytesReadToDisk += other.remoteBytesReadToDisk;
            this.localBytesRead += other.localBytesRead;
            this.recordsRead += other.recordsRead;
            this.remoteReqsDuration += other.remoteReqsDuration;
            if (other.shufflePushReadMetrics != null) {
                if (this.shufflePushReadMetrics == null) {
                    this.shufflePushReadMetrics = new ShufflePushReadMetrics();
                }
                this.shufflePushReadMetrics.incrMetrics(other.shufflePushReadMetrics);
            }
        }
    }


    @NoArgsConstructor @AllArgsConstructor
    public static class ShufflePushReadMetrics {

        @JsonProperty("Corrupt Merged Block Chunks")
        public long corruptMergedBlockChunks;

        @JsonProperty("Merged Fetch Fallback Count")
        public long mergedFetchFallbackCount;

        @JsonProperty("Merged Remote Blocks Fetched")
        public long remoteMergedBlocksFetched;

        @JsonProperty("Merged Local Blocks Fetched")
        public long localMergedBlocksFetched;

        @JsonProperty("Merged Remote Chunks Fetched")
        public long remoteMergedChunksFetched;

        @JsonProperty("Merged Local Chunks Fetched")
        public long localMergedChunksFetched;

        @JsonProperty("Merged Remote Bytes Read")
        public long remoteMergedBytesRead;

        @JsonProperty("Merged Local Bytes Read")
        public long localMergedBytesRead;

        @JsonProperty("Merged Remote Requests Duration")
        public long remoteMergedReqsDuration;

        public void incrMetrics(ShufflePushReadMetrics other) {
            this.corruptMergedBlockChunks += other.corruptMergedBlockChunks;
            this.mergedFetchFallbackCount += other.mergedFetchFallbackCount;
            this.remoteMergedBlocksFetched += other.remoteMergedBlocksFetched;
            this.localMergedBlocksFetched += other.localMergedBlocksFetched;
            this.remoteMergedChunksFetched += other.remoteMergedChunksFetched;
            this.localMergedChunksFetched += other.localMergedChunksFetched;
            this.remoteMergedBytesRead += other.remoteMergedBytesRead;
            this.localMergedBytesRead += other.localMergedBytesRead;
            this.remoteMergedReqsDuration += other.remoteMergedReqsDuration;
        }
    }

    @NoArgsConstructor @AllArgsConstructor
    public static class ShuffleWriteMetrics {
    	@JsonProperty("Shuffle Bytes Written")
	    public long bytesWritten;
	    
	    @JsonProperty("Shuffle Write Time")
	    public long writeTime;

	    @JsonProperty("Shuffle Records Written")
	    public long recordsWritten;

        public void incrMetrics(ShuffleWriteMetrics other) {
            this.bytesWritten += other.bytesWritten;
            this.writeTime += other.writeTime;
            this.recordsWritten += other.recordsWritten;
        }
    }

    @NoArgsConstructor @AllArgsConstructor
    public static class BlockIdStatus {
        @JsonProperty("Block ID")
        public String blockId;

        @JsonProperty("Status")
        public String status;
    }





    /**
     * Construct a [[TaskMetrics]] object from a list of [[AccumulableInfo]], called on driver only.
     * The returned [[TaskMetrics]] is only used to get some internal metrics, we don't need to take
     * care of external accumulator info passed in.
     */
    public static TaskMetrics fromAccumulatorInfos(Collection<AccumulableInfo> infos) {
        val tm = new TaskMetrics();
        for(val info : infos) {
            if (info.name != null && info.update != null) {
                val name = info.name;
                val value = info.update;
                if (name.equals(UPDATED_BLOCK_STATUSES)) {
                    @SuppressWarnings({"unchecked", "rawtypes"})
                    val valueLs = (List<BlockIdStatus>) (List) value;
                    tm.setUpdatedBlockStatuses(valueLs);
                } else {
                    // TOADD accumulators on spark-side only ?
                    // tm.nameToAccums.get(name).foreach(
                    //        _.asInstanceOf[LongAccumulator].setValue(value.asInstanceOf[Long])
                    // )
                }
            }
        }
        return tm;
    }


//    /**
//     * Construct a [[TaskMetrics]] object from a list of accumulator updates, called on driver only.
//     */
//    def fromAccumulators(accums: Seq[AccumulatorV2[_, _]]): TaskMetrics = {
//        val tm = new TaskMetrics
//        for (acc <- accums) {
//            val name = acc.name
//            if (name.isDefined && tm.nameToAccums.contains(name.get)) {
//                val tmAcc = tm.nameToAccums(name.get).asInstanceOf[AccumulatorV2[Any, Any]]
//                tmAcc.metadata = acc.metadata
//                tmAcc.merge(acc.asInstanceOf[AccumulatorV2[Any, Any]])
//            } else {
//                tm._externalAccums += acc
//            }
//        }
//        tm
//    }



}
